### PR TITLE
Update traj data generation code

### PR DESCRIPTION
**Pull request by JinraeKim**
#54 (partially resolved)

Merge it after #53 is reviewed and merged.

### 분석
- #53 와 비교했을 때, test loss 의 오더가 상승함
    - 데이터 개수를 늘렸을 때 이것이 개선될지는 확인이 필요
- cost 값이 대충 0~2 정도 분포했음. 즉, 최대값 대비 평균 2.5% 정도의 오차를 보임 -> 데이터 개수를 500개로 늘리니 0.5%정도의 오차로 줄어듦

### 결과 1
- 초기조건이 고정되고, 궤적 파라미터 (상수 궤적) 를 랜덤하게 변경한 100개의 데이터로 학습함
```julia
epoch: 490/500
train loss: 0.04503805318259698, test loss: 0.04810623734922672
epoch: 491/500
train loss: 0.04430858432891502, test loss: 0.047071281567201056
epoch: 492/500
train loss: 0.045794739081250216, test loss: 0.04412293497425684
epoch: 493/500
train loss: 0.044308797635761635, test loss: 0.04541345746295398
epoch: 494/500
train loss: 0.04509754495954448, test loss: 0.052094933080495086
epoch: 495/500
train loss: 0.0444691486654015, test loss: 0.05142818097239678
epoch: 496/500
train loss: 0.0445010070274344, test loss: 0.0476656591470513
epoch: 497/500
train loss: 0.04435291862398784, test loss: 0.04621449524361766
epoch: 498/500
train loss: 0.04404660949320234, test loss: 0.04525638869335364
epoch: 499/500
train loss: 0.04405194051662392, test loss: 0.046091424930600464
epoch: 500/500
train loss: 0.04400790758492827, test loss: 0.04486687605259744
```
### 결과 2
- 초기조건이 고정되고, 궤적 파라미터 (상수 궤적) 를 랜덤하게 변경한 500개의 데이터로 학습함. epoch를 1000개로 늘림
```julia
epoch: 990/1000
train loss: 0.006585062535652302, test loss: 0.009129077215332525
epoch: 991/1000
train loss: 0.00590244877598379, test loss: 0.009625629937753262
epoch: 992/1000
train loss: 0.0072928134854061575, test loss: 0.011373839406195225
epoch: 993/1000
train loss: 0.006306521163276164, test loss: 0.010751361105481338
epoch: 994/1000
train loss: 0.017537169693366897, test loss: 0.02009221965855168
epoch: 995/1000
train loss: 0.0061754190715973874, test loss: 0.008788245272527154
epoch: 996/1000
train loss: 0.006484239708325825, test loss: 0.009061958436653713
epoch: 997/1000
train loss: 0.00666492314316285, test loss: 0.010788371089625439
epoch: 998/1000
train loss: 0.006822148854414358, test loss: 0.009971723669883005
epoch: 999/1000
train loss: 0.008385256711416616, test loss: 0.012257413496356314
epoch: 1000/1000
train loss: 0.006311904495964465, test loss: 0.009735567733803761

```

### 네트워크 구조
- single hidden layer feedforward neural network
```julia
    n_h = 256  # hidden layer nodes
    Ĵ = Chain(
              Dense(n, n_h, leakyrelu),
              Dense(n_h, 1, leakyrelu),
             )
```

### 데이터 읽고 쓰기
- preprocessing 과정에서 시간이 꽤 소요됨 (100개: 약 10초 내외, 500개: 약 40초 내외). 이를 개선하려면, preprocessing 된 데이터를 따로 저장해두고 불러와서 쓰는게 좋을듯
### 하이퍼 파라미터
- 최고의 성능을 보이기 위해서는 약간의 튜닝이 필요했음 (네트워크 구조 등)
- mini-batch size 가 너무 작으면 오버피팅이 발생함. 32 이상에서 해당 현상이 완화됨을 확인함
### 초기조건도 cost approximator 의 변수에 추가한다면?
- 현재는 상태변수 차원이 18차원. quaternion 변환을 한다면 13차원까지 줄일 수 있을 것으로 보임 (단위 쿼터니언이므로 대충 12차원).
    - 간단하게 로컬에서 18차원 상태변수를 추가했을 때, test loss 값이 꽤 큰 것을 확인함 (1정도; 그러나 최대 cost 도 더 컸음). 이는 다른 PR 을 통해 개선 또는 기능 추가해야할 것으로 생각됨

### CODE REVIEW

**Comment by JinraeKim** [comment]
`main` is rebased

**Comment by JinraeKim** [comment]
Note: 네트워크 구조를 아래와 같이 바꾸면,
```julia
    n_h = 128  # hidden layer nodes
    Ĵ = Chain(
              Dense(n, n_h, leakyrelu),
              Dense(n_h, n_h, leakyrelu),
              Dense(n_h, 1, leakyrelu),
             )
```

평균 오차가 0.1% 까지 줄어듦을 확인; 성능은 충분하다고 판단되어, 더이상의 튜닝은 진행하지 않겠음
```julia
epoch: 990/1000
train loss: 0.00027136580426713805, test loss: 0.001926573600667545
epoch: 991/1000
train loss: 0.0002851237678318677, test loss: 0.0018308164143443712
epoch: 992/1000
train loss: 0.00031937771302380995, test loss: 0.0021154499678588006
epoch: 993/1000
train loss: 0.00027319437320415143, test loss: 0.0016502549024506683
epoch: 994/1000
train loss: 0.0005545717759468194, test loss: 0.0022142406340577908
epoch: 995/1000
train loss: 0.000704409573516152, test loss: 0.001446155151335877
epoch: 996/1000
train loss: 0.000406484253303953, test loss: 0.0021487459985304854
epoch: 997/1000
train loss: 0.0004334657128200384, test loss: 0.0015424979102669895
epoch: 998/1000
train loss: 0.0004618592443763306, test loss: 0.0021054212515375847
epoch: 999/1000
train loss: 0.00019264890760698394, test loss: 0.0016927620567668722
epoch: 1000/1000
train loss: 0.0002958085313238846, test loss: 0.0022949433634259055
```

**Comment by hnlee77** [comment]
> ### 초기조건도 cost approximator 의 변수에 추가한다면?
> * 현재는 상태변수 차원이 18차원. quaternion 변환을 한다면 13차원까지 줄일 수 있을 것으로 보임 (단위 쿼터니언이므로 대충 12차원).
>   
>   * 간단하게 로컬에서 18차원 상태변수를 추가했을 때, test loss 값이 꽤 큰 것을 확인함 (1정도; 그러나 최대 cost 도 더 컸음). 이는 다른 PR 을 통해 개선 또는 기능 추가해야할 것으로 생각됨

이것은 확실히 쿼터니언을 이용해 차원을 줄일 필요가 있겠군요. test loss값이 성능이 괜찮은 것으로 보이네요!

**Comment by hnlee77** [comment]
이거 완성된 PR맞나요?

**Comment by JinraeKim** [comment]
> 이거 완성된 PR맞나요?

왜 그렇게 생각하는지를 말해주실 수 있나요?

물론, 아래 내용 또한 참고 바람

> #54 (partially resolved)
> 
> Merge it after #53 is reviewed and merged.

**Comment by hnlee77** [comment]
> > 이거 완성된 PR맞나요?
> 
> 왜 그렇게 생각하는지를 말해주실 수 있나요?

아 별건 아니고.. Review requested 가 되어있지 않길래 물어봤어요 ㅎㅎ 하필 마지막 커밋 이름도 `wip`여서 ㅎㅎ

**Comment by JinraeKim** [comment]
> > > 이거 완성된 PR맞나요?
> > 
> > 
> > 왜 그렇게 생각하는지를 말해주실 수 있나요?
> 
> 아 별건 아니고.. Review requested 가 되어있지 않길래 물어봤어요 ㅎㅎ 하필 마지막 커밋 이름도 `wip`여서 ㅎㅎ

죄송합니다 ㅎㅎ;

**Review by hnlee77 — APPROVED** [review]

**Comment by JinraeKim** [comment]
#53 rebased
